src/components/network/line-layer.js

```diff
@@ -77,11 +111,39 @@ class LineLayer extends CompositeLayer {
 
             this.setState({compositeData: compositeData});
         }
+        //update the label position when toggling fullPath
+        if (changeFlags.propsChanged) {
```

Comment:
Please change to have only one code that initializes the positions/distances and active power labels. Maybe use (if changeFlags.propsOrDataChanged) or something else ?

Reply:
I don't see how this would work, I don't need to do the same processing. When the page first load we need to make several initialization, basically populate the compositeData array. However, when the prop fullPath changes i need to update only the label position and offset. The propsOrDataChanged wouldn't help as I would just do inside a if (propsChanged) inside as i need to do that only when one specific prop changed.

Reply:
let's do this in another PR

Reply:
This code is almost duplicated and could be unified :
```
                compositeData.forEach(compositeData => {
                    let lineMap = new Map();
                    compositeData.lines.forEach(line => {
                        const positions = props.geoData.getLinePositions(props.network, line, props.lineFullPath);
                        const cumulativeDistances = props.geoData.getLineDistances(positions);
                        lineMap.set(line.id, {positions: positions, cumulativeDistances: cumulativeDistances, line: line});
                    });
                    compositeData.lineMap = lineMap;
                });
``` 
and
```
                    compositeData.lineMap.forEach(lineData => {
                        lineData.positions = props.geoData.getLinePositions(props.network, lineData.line, props.lineFullPath);
                        lineData.cumulativeDistances = props.geoData.getLineDistances(lineData.positions);
                    });
```

same for
```
                        let coordinates1 = props.geoData.getCoordinateInLine(lineData.positions, lineData.cumulativeDistances, 15);
                        let coordinates2 = props.geoData.getCoordinateInLine(lineData.positions, lineData.cumulativeDistances, 85);
                        if (coordinates1 !== null && coordinates2 !== null) {
                            compositeData.activePower.push({
                                line: line,
                                p: line.p1,
                                percent: 15,
                                printPosition: [coordinates1.distance.longitude, coordinates1.distance.latitude],
                                offset: coordinates1.offset,
                                });
                            compositeData.activePower.push({
                                line: line,
                                p: line.p2,
                                percent: 85,
                                printPosition: [coordinates2.distance.longitude, coordinates2.distance.latitude],
                                offset: coordinates2.offset});
                        }
                        line.cumulativeDistances = lineData.cumulativeDistances;
                        line.positions = lineData.positions;
```
and
```
                    compositeData.activePower.forEach(activePower => {
                        let lineData = compositeData.lineMap.get(activePower.line.id);
                        const positions = lineData.positions;
                        const cumulativeDistances = lineData.cumulativeDistances;
                        let coordinates = props.geoData.getCoordinateInLine(positions, cumulativeDistances, activePower.percent);
                        if (coordinates) {
                            activePower.printPosition = [coordinates.distance.longitude, coordinates.distance.latitude];
                            activePower.offset = coordinates.offset;
                        }
                    });
                    compositeData.arrows.forEach(arrows => {
                        let lineData = compositeData.lineMap.get(arrows.line.id);
                        arrows.line.positions = lineData.positions;
                        arrows.line.cumulativeDistances = lineData.cumulativeDistances;
                    });
```

it should all go in something like this:
```
if (propsOrDataChanged) {
  if (dataChanged || (propsChanged && oldProps.lineFullPath !== undefined && props.lineFullPath) {
     //the refactored code here
  }
}
```
no ?

Reply:
`compositeData.forEach(compositeData => {
                    let lineMap = new Map();
                    compositeData.lines.forEach(line => {
                        const positions = props.geoData.getLinePositions(props.network, line, props.lineFullPath);
                        const cumulativeDistances = props.geoData.getLineDistances(positions);
                        lineMap.set(line.id, {positions: positions, cumulativeDistances: cumulativeDistances, line: line});
                    });
                    compositeData.lineMap = lineMap;
                });`
here I just initialize the structure

`compositeData.lineMap.forEach(lineData => {
                        lineData.positions = props.geoData.getLinePositions(props.network, lineData.line, props.lineFullPath);
                        lineData.cumulativeDistances = props.geoData.getLineDistances(lineData.positions);
                    });`
and there I update some fields of it.

and same for the 2 others. I don't understand how or why we should try to merge them into 1 code, they don't serve the same purpose and are not called at the same time. I even find it better to have distinct part with clear goal.

Reply:
Ok so we accept to recreate everything when the prop fullLine changes to have only one code